revocation/crl/cache/file.go

```diff
@@ -0,0 +1,272 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"archive/tar"
+	"context"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+const (
+	// tempFileName is the prefix of the temporary file
+	tempFileName = "notation-*"
+)
+
+// FileCache stores in a tarball format, which contains two files: base.crl and
+// metadata.json. The base.crl file contains the base CRL in DER format, and the
+// metadata.json file contains the metadata of the CRL.
+//
+// The cache builds on top of the UNIX file system to leverage the file system's
+// atomic operations. The `rename` and `remove` operations will unlink the old
+// file but keep the inode and file descriptor for existing processes to access
+// the file. The old inode will be dereferenced when all processes close the old
+// file descriptor. Additionally, the operations are proven to be atomic on
+// UNIX-like platforms, so there is no need to handle file locking.
+//
+// NOTE: For Windows, the `open`, `rename` and `remove` operations need file
+// locking to ensure atomicity. The current implementation does not handle
+// file locking, so the concurrent write from multiple processes may be failed.
+// Please do not use this cache in a multi-process environment on Windows.
+//
+// FileCache doesn't handle cache cleaning but provides the Delete and Clear
+// methods to remove the CRLs from the file system.
+type FileCache struct {
+	// MaxAge is the maximum age of the CRLs cache. If the CRL is older than
+	// MaxAge, it will be considered as expired.
+	MaxAge time.Duration
+
+	root string
+}
+
+// NewFileCache creates a new file system store
+//
+//   - root is the directory to store the CRLs.
+func NewFileCache(root string) (*FileCache, error) {
+	if err := os.MkdirAll(root, 0700); err != nil {
+		return nil, fmt.Errorf("failed to create directory: %w", err)
+	}
+
+	return &FileCache{
+		MaxAge: DefaultMaxAge,
+		root:   root,
+	}, nil
+}
+
+// Get retrieves the CRL bundle from the file system
+//
+// - if the key does not exist, return ErrNotFound
+// - if the CRL is expired, return ErrCacheMiss
+func (c *FileCache) Get(ctx context.Context, uri string) (bundle *Bundle, err error) {
+	f, err := os.Open(filepath.Join(c.root, fileName(uri)))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrNotFound
```

Comment:
Is it an `ErrCacheMiss` if the file is not found?

Reply:
Updated.